debian/mongorestore.1
.\" Man page generated from reStructuredText.
.
.TH "MONGORESTORE" "1" "Oct 24, 2019" "4.0" "mongodb-manual"
.SH NAME
mongorestore \- MongoDB Data Restoration Tool
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Syntax\fP
.IP \(bu 2
\fI\%Behavior\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Examples\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "macOS Sierra and Go 1.6 Incompatibility"
.sp
Users running on macOS Sierra require the 3.2.10 or newer version
of  mongorestore\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
The \fI\%mongorestore\fP program loads data from either a binary
database dump created by \fBmongodump\fP or the standard input
(starting in version 3.0.0) into a \fBmongod\fP or
\fBmongos\fP instance.
.sp
For an overview of \fI\%mongorestore\fP usage, see
/tutorial/backup\-and\-restore\-tools\&.
.SH SYNTAX
.sp
Run \fI\%mongorestore\fP from the system command line, not the \fBmongo\fP shell.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongorestore [options] [<directory>/<BSON file>]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For example, to restore from a \fBdump\fP directory to a local
\fBmongod\fP instance running on port \fB27017\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongorestore  dump/
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
As \fI\%mongorestore\fP restores from the \fBdump/\fP directory,
it creates the database and collections as needed and logs its progress:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-07\-13T10:05:40.353\-0400   preparing collections to restore from
2019\-07\-13T10:05:40.386\-0400   reading metadata for test.bakesales from dump/test/bakesales.metadata.json
2019\-07\-13T10:05:40.403\-0400   reading metadata for test.salaries from dump/test/salaries.metadata.json
2019\-07\-13T10:05:40.434\-0400   restoring test.bakesales from dump/test/bakesales.bson
2019\-07\-13T10:05:40.464\-0400   restoring test.salaries from dump/test/salaries.bson
2019\-07\-13T10:05:40.473\-0400   no indexes to restore
2019\-07\-13T10:05:40.473\-0400   finished restoring test.bakesales (21 documents)
2019\-07\-13T10:05:40.490\-0400   no indexes to restore
2019\-07\-13T10:05:40.491\-0400   finished restoring test.salaries (10 documents)
2019\-07\-13T10:05:40.491\-0400   restoring users from dump/admin/system.users.bson
2019\-07\-13T10:05:40.559\-0400   restoring roles from dump/admin/system.roles.bson
2019\-07\-13T10:05:40.711\-0400   done
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can also restore a specific collection or collections from the
\fBdump/\fP directory. For example, the following operation restores a
single collection from corresponding data files in the \fBdump/\fP
directory:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongorestore \-\-nsInclude test.purchaseorders dump/
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If the \fBdump/\fP directory does not contain the corresponding data file
for the specified namespace, no data will be restored. For example, the
following specifies a collection namespace that does not have a
corresponding data in the \fBdump/\fP directory:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongorestore \-\-nsInclude foo.bar dump/
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \fI\%mongorestore\fP outputs the following messages:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-07\-13T10:06:36.095\-0400   preparing collections to restore from
2019\-07\-13T10:06:36.095\-0400   done
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For more examples, see \fI\%Examples\fP\&.
.sp
For more information on the options and arguments, see
\fI\%Options\fP\&.
.SH BEHAVIOR
.SS Insert Only
.sp
\fI\%mongorestore\fP can create a new database or add data to an
existing database. However, \fI\%mongorestore\fP performs inserts
only and does not perform updates. That is, if restoring documents to
an existing database and collection and existing documents have the
same value \fB_id\fP field as the to\-be\-restored documents,
\fI\%mongorestore\fP will \fInot\fP overwrite those documents.
.SS Rebuild Indexes
.sp
\fI\%mongorestore\fP recreates indexes recorded by
\fBmongodump\fP\&.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
Starting in MongoDB 2.6, creating indexes will error if an
index key in an existing document exceeds the limit\&. See
2.6\-index\-key\-length\-incompatibility for more information and
solution.
.sp
If you have an existing data set that violates this limit but want
to resolve the index issue after restoring the data, you can disable
the default index key length validation on the target database by
setting the \fBmongod\fP instance\(aqs
\fBfailIndexKeyTooLong\fP parameter to false.
.UNINDENT
.UNINDENT
.SS Version Compatibility
.sp
The data format used by \fBmongodump\fP from version 2.2 or
later is \fIincompatible\fP with earlier versions of \fBmongod\fP\&.
Do not use recent versions of \fBmongodump\fP to back up older
data stores.
.SS Exclude \fBsystem.profile\fP Collection
.sp
\fI\%mongorestore\fP does not restore the \fBsystem.profile\fP collection data.
.SH REQUIRED ACCESS
.sp
To restore data to a MongoDB deployment that has access control enabled, the \fBrestore\fP role provides
the necessary privileges to restore data from backups \fIif\fP the data does
not include \fBsystem.profile\fP
collection data and you run \fI\%mongorestore\fP without the
\fI\%\-\-oplogReplay\fP option.
.sp
If the backup data includes \fBsystem.profile\fP collection data or you run with
\fI\%\-\-oplogReplay\fP, you need
additional privileges:
.TS
center;
|l|l|.
_
T{
\fBsystem.profile\fP
T}	T{
If the backup data includes \fBsystem.profile\fP collection data and the target database
does not contain the \fBsystem.profile\fP
collection, \fI\%mongorestore\fP attempts to create the collection
even though the program does not actually restore \fBsystem.profile\fP
documents. As such, the user requires additional privileges to perform
\fBcreateCollection\fP and \fBconvertToCapped\fP
actions on the \fBsystem.profile\fP
collection for a database.
.sp
Both the built\-in roles \fBdbAdmin\fP and
\fBdbAdminAnyDatabase\fP provide the additional privileges.
T}
_
T{
\fB\-\-oplogReplay\fP
T}	T{
To run with \fI\%\-\-oplogReplay\fP, create a user\-defined role that has \fBanyAction\fP on
resource\-anyresource\&.
.sp
Grant only to users who must run \fI\%mongorestore\fP
with \fI\%\-\-oplogReplay\fP\&.
T}
_
.TE
.SH OPTIONS
.sp
Changed in version 3.0.0: \fI\%mongorestore\fP removed the \fB\-\-filter\fP, \fB\-\-dbpath\fP, and the
\fB\-\-noobjcheck\fP options.

.INDENT 0.0
.TP
.B mongorestore
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongorestore\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBmongorestore\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongorestore\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI connection string to connect to the MongoDB deployment.
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-uri "mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For more information on the components of the connection string, see
the Connection String URI Format documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following command\-line options cannot be used in conjunction
with \fI\%\-\-uri\fP option:
.INDENT 0.0
.IP \(bu 2
\fI\%\-\-host\fP
.IP \(bu 2
\fI\%\-\-port\fP
.IP \(bu 2
\fI\%\-\-db\fP
.IP \(bu 2
\fI\%\-\-username\fP
.IP \(bu 2
\fI\%\-\-password\fP  (if the
URI connection string also includes the password)
.IP \(bu 2
\fI\%\-\-authenticationDatabase\fP
.IP \(bu 2

NEWS
Ticket reference tags refer to tickets in the MongoDB JIRA tracking system:
https://jira.mongodb.org

WiredTiger release 2.9.3, 2017-06-27
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* WT-2972       Add an interface allowing partial updates to existing values
* WT-3063       Add an interface allowing reservation of records for read-modify-write
* WT-3142       Add a workload generator application
* WT-3160       Improve eviction of internal pages from idle trees
* WT-3245       Avoid hangs on shutdown when a utility thread encounters an error
* WT-3258       Improve visibility into thread wait time due to pages exceeding memory_page_max
* WT-3263       Allow archive on restart/recovery if clean shutdown
* WT-3287       Review WiredTiger internal panic checks
* WT-3292       Review/cleanup full-barrier calls in WiredTiger
* WT-3296       LAS table fixes/improvements
* WT-3327       Checkpoints can hang if time runs backward
* WT-3345       Improve rwlock scaling
* WT-3373       Access violation due to a bug in internal page splitting
* WT-3379       Change when pages can be split to avoid excessively slowing some operations

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/18291

WiredTiger release 2.9.2, 2017-05-25
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* SERVER-16796  Increase logging activity for journal recovery operations
* SERVER-28168  Cannot start or repair MongoDB after unexpected shutdown.
* SERVER-28194  Fix a bug where a missing WiredTiger.turtle file can result in data loss
* WT-98         Update the current cursor value without a search
* WT-2439       Enhance reconciliation page layout
* WT-2898       Improve performance of eviction-heavy workloads by dynamically controlling the number of eviction threads
* WT-3097       Race on reconfigure or shutdown can lead to waiting for statistics log server
* WT-3106       Add truncate support to command line wt utility
* WT-3114       Avoid archiving log files immediately after recovery
* WT-3115       Change the dhandle lock to a read/write lock
* WT-3127       Fix a bug with CPU yield calls don't necessarily imply memory barriers
* WT-3135       Fix a bug with search_near() for index with custom collator
* WT-3137       Fix a hang in __log_slot_join/__log_slot_switch_internal
* WT-3144       Fix a bug where random cursor returns not-found when descending to an empty page
* WT-3148       Improve eviction efficiency with many small trees
* WT-3149       Change eviction to start new walks from a random place in the tree
* WT-3150       Reduce impact of checkpoints on eviction server
* WT-3152       Convert table lock from a spinlock to a read write lock
* WT-3157       Fix a bug in checkpoint/transaction integrity issue when writes fail.
* WT-3158       Fix structure packing on Windows.
* WT-3188       Fix error handling in logging where fatal errors could lead to a hang
* WT-3193       Close a race between verify opening a handle and eviction visiting it
* WT-3206       Fix a bug where core dump happened on NULL page index
* WT-3218       Fix a bug that could lead to unexpected checkpoint ordering failures
* WT-3243       Reorder log slot release so joins don't wait on IO
* WT-3262       Stop making schema operations wait for cache

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/17874


WiredTiger release 2.9.1, 2016-12-22
------------------------------------

New features and API changes; refer to the API documentation for full details:

* SERVER-26545	Remove fixed-size limitation on WiredTiger hazard pointers. See the upgrading documentation for details
* WT-283	Add a new WT_SESSION::alter method that can be used to reconfigure table metadata
* WT-2670	Change the default file system access pattern advice for data files from random to no advice. Add access_pattern_hint configuration option for WT_SESSION::create API that can be used to advise the file system of expected access semantics. See the upgrading documentation for details.
* WT-3034	Add support for including updates when reading from named snapshots

Significant changes and bug fixes:


src/mongo/gotools/mongorestore/mongorestore.go
// Copyright (C) MongoDB, Inc. 2014-present.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License. You may obtain
// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

// Package mongorestore writes BSON data to a MongoDB instance.
package mongorestore

import (
	"compress/gzip"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"sync"

	"github.com/mongodb/mongo-tools/common/archive"
	"github.com/mongodb/mongo-tools/common/auth"
	"github.com/mongodb/mongo-tools/common/db"
	"github.com/mongodb/mongo-tools/common/intents"
	"github.com/mongodb/mongo-tools/common/log"
	"github.com/mongodb/mongo-tools/common/options"
	"github.com/mongodb/mongo-tools/common/progress"
	"github.com/mongodb/mongo-tools/common/util"
	"github.com/mongodb/mongo-tools/mongorestore/ns"
	"gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
)

// MongoRestore is a container for the user-specified options and
// internal state used for running mongorestore.
type MongoRestore struct {
	ToolOptions   *options.ToolOptions
	InputOptions  *InputOptions
	OutputOptions *OutputOptions
	NSOptions     *NSOptions

	SessionProvider *db.SessionProvider
	ProgressManager progress.Manager

	TargetDirectory string

	// Skip restoring users and roles, regardless of namespace, when true.
	SkipUsersAndRoles bool

	// other internal state
	manager *intents.Manager
	safety  *mgo.Safe

	objCheck         bool
	oplogLimit       bson.MongoTimestamp
	isMongos         bool
	useWriteCommands bool
	authVersions     authVersionPair

	// a map of database names to a list of collection names
	knownCollections      map[string][]string
	knownCollectionsMutex sync.Mutex

	renamer  *ns.Renamer
	includer *ns.Matcher
	excluder *ns.Matcher

	// indexes belonging to dbs and collections
	dbCollectionIndexes map[string]collectionIndexes

	archive *archive.Reader

	// channel on which to notify if/when a termination signal is received
	termChan chan struct{}

	// Reader to take care of BSON input if not reading from the local filesystem.
	// This is initialized to os.Stdin if unset.
	InputReader io.Reader
}

type collectionIndexes map[string][]IndexDocument

// ParseAndValidateOptions returns a non-nil error if user-supplied options are invalid.
func (restore *MongoRestore) ParseAndValidateOptions() error {
	// Can't use option pkg defaults for --objcheck because it's two separate flags,
	// and we need to be able to see if they're both being used. We default to
	// true here and then see if noobjcheck is enabled.
	log.Logv(log.DebugHigh, "checking options")
	if restore.InputOptions.Objcheck {
		restore.objCheck = true
		log.Logv(log.DebugHigh, "\tdumping with object check enabled")
	} else {
		log.Logv(log.DebugHigh, "\tdumping with object check disabled")
	}

	if restore.NSOptions.DB == "" && restore.NSOptions.Collection != "" {
		return fmt.Errorf("cannot restore a collection without a specified database")
	}

	if restore.NSOptions.DB != "" {
		if err := util.ValidateDBName(restore.NSOptions.DB); err != nil {
			return fmt.Errorf("invalid db name: %v", err)
		}
	}
	if restore.NSOptions.Collection != "" {
		if err := util.ValidateCollectionGrammar(restore.NSOptions.Collection); err != nil {
			return fmt.Errorf("invalid collection name: %v", err)
		}
	}
	if restore.InputOptions.RestoreDBUsersAndRoles && restore.NSOptions.DB == "" {
		return fmt.Errorf("cannot use --restoreDbUsersAndRoles without a specified database")
	}
	if restore.InputOptions.RestoreDBUsersAndRoles && restore.NSOptions.DB == "admin" {
		return fmt.Errorf("cannot use --restoreDbUsersAndRoles with the admin database")
	}

	var err error
	restore.isMongos, err = restore.SessionProvider.IsMongos()
	if err != nil {
		return err
	}
	if restore.isMongos {
		log.Logv(log.DebugLow, "restoring to a sharded system")
	}

	if restore.InputOptions.OplogLimit != "" {
		if !restore.InputOptions.OplogReplay {
			return fmt.Errorf("cannot use --oplogLimit without --oplogReplay enabled")
		}
		restore.oplogLimit, err = ParseTimestampFlag(restore.InputOptions.OplogLimit)
		if err != nil {
			return fmt.Errorf("error parsing timestamp argument to --oplogLimit: %v", err)
		}
	}
	if restore.InputOptions.OplogFile != "" {
		if !restore.InputOptions.OplogReplay {
			return fmt.Errorf("cannot use --oplogFile without --oplogReplay enabled")
		}
		if restore.InputOptions.Archive != "" {
			return fmt.Errorf("cannot use --oplogFile with --archive specified")
		}
	}

	// check if we are using a replica set and fall back to w=1 if we aren't (for <= 2.4)
	nodeType, err := restore.SessionProvider.GetNodeType()
	if err != nil {
		return fmt.Errorf("error determining type of connected node: %v", err)
	}

	log.Logvf(log.DebugLow, "connected to node type: %v", nodeType)
	restore.safety, err = db.BuildWriteConcern(restore.OutputOptions.WriteConcern, nodeType,
		restore.ToolOptions.URI.ParsedConnString())
	if err != nil {
		return fmt.Errorf("error parsing write concern: %v", err)
	}

	// deprecations with --nsInclude --nsExclude
	if restore.NSOptions.DB != "" || restore.NSOptions.Collection != "" {
		// these are only okay if restoring from a bson file
		_, fileType := restore.getInfoFromFilename(restore.TargetDirectory)
		if fileType != BSONFileType {
			log.Logvf(log.Always, "the --db and --collection args should only be used when "+
				"restoring from a BSON file. Other uses are deprecated and will not exist "+
				"in the future; use --nsInclude instead")
		}
	}
	if len(restore.NSOptions.ExcludedCollections) > 0 ||
		len(restore.NSOptions.ExcludedCollectionPrefixes) > 0 {
		log.Logvf(log.Always, "the --excludeCollections and --excludeCollectionPrefixes options "+
			"are deprecated and will not exist in the future; use --nsExclude instead")
	}
	if restore.InputOptions.OplogReplay {
		if len(restore.NSOptions.NSInclude) > 0 || restore.NSOptions.DB != "" {
			return fmt.Errorf("cannot use --oplogReplay with includes specified")
		}
		if len(restore.NSOptions.NSExclude) > 0 || len(restore.NSOptions.ExcludedCollections) > 0 ||
			len(restore.NSOptions.ExcludedCollectionPrefixes) > 0 {
			return fmt.Errorf("cannot use --oplogReplay with excludes specified")
		}
		if len(restore.NSOptions.NSFrom) > 0 {
			return fmt.Errorf("cannot use --oplogReplay with namespace renames specified")
		}
	}

	includes := restore.NSOptions.NSInclude
	if restore.NSOptions.DB != "" && restore.NSOptions.Collection != "" {
		includes = append(includes, ns.Escape(restore.NSOptions.DB)+"."+
			restore.NSOptions.Collection)
	} else if restore.NSOptions.DB != "" {
		includes = append(includes, ns.Escape(restore.NSOptions.DB)+".*")
	}
	if len(includes) == 0 {
		includes = []string{"*"}
	}
	restore.includer, err = ns.NewMatcher(includes)
	if err != nil {
		return fmt.Errorf("invalid includes: %v", err)
	}

	if len(restore.NSOptions.ExcludedCollections) > 0 && restore.NSOptions.Collection != "" {
		return fmt.Errorf("--collection is not allowed when --excludeCollection is specified")
	}
	if len(restore.NSOptions.ExcludedCollectionPrefixes) > 0 && restore.NSOptions.Collection != "" {
		return fmt.Errorf("--collection is not allowed when --excludeCollectionsWithPrefix is specified")
	}
	excludes := restore.NSOptions.NSExclude
	for _, col := range restore.NSOptions.ExcludedCollections {
		excludes = append(excludes, "*."+ns.Escape(col))
	}
	for _, colPrefix := range restore.NSOptions.ExcludedCollectionPrefixes {
		excludes = append(excludes, "*."+ns.Escape(colPrefix)+"*")
	}
	restore.excluder, err = ns.NewMatcher(excludes)
	if err != nil {
		return fmt.Errorf("invalid excludes: %v", err)
	}

	if len(restore.NSOptions.NSFrom) != len(restore.NSOptions.NSTo) {
		return fmt.Errorf("--nsFrom and --nsTo arguments must be specified an equal number of times")
	}
	restore.renamer, err = ns.NewRenamer(restore.NSOptions.NSFrom, restore.NSOptions.NSTo)
	if err != nil {
		return fmt.Errorf("invalid renames: %v", err)
	}

	if restore.OutputOptions.NumInsertionWorkers < 0 {
		return fmt.Errorf(
			"cannot specify a negative number of insertion workers per collection")
	}

	if restore.OutputOptions.PreserveUUID {
		if !restore.OutputOptions.Drop {
			return fmt.Errorf("cannot specify --preserveUUID without --drop")
		}

		ok, err := restore.SessionProvider.SupportsCollectionUUID()
		if err != nil {
			return err
		}
		if !ok {
			return fmt.Errorf("target host does not support --preserveUUID")
		}
	}

	// a single dash signals reading from stdin
	if restore.TargetDirectory == "-" {
		if restore.InputOptions.Archive != "" {
			return fmt.Errorf(
				"cannot restore from \"-\" when --archive is specified")
		}
		if restore.NSOptions.Collection == "" {
			return fmt.Errorf("cannot restore from stdin without a specified collection")
		}
	}
	if restore.InputReader == nil {
		restore.InputReader = os.Stdin
	}

	return nil
}

// Restore runs the mongorestore program.
func (restore *MongoRestore) Restore() error {
	var target archive.DirLike
	err := restore.ParseAndValidateOptions()
	if err != nil {
		log.Logvf(log.DebugLow, "got error from options parsing: %v", err)
		return err
	}

	// Build up all intents to be restored
	restore.manager = intents.NewIntentManager()
	if restore.InputOptions.Archive == "" && restore.InputOptions.OplogReplay {
		restore.manager.SetSmartPickOplog(true)
	}

	if restore.InputOptions.Archive != "" {
		if restore.archive == nil {
			archiveReader, err := restore.getArchiveReader()
			if err != nil {
				return err
			}
			restore.archive = &archive.Reader{
				In:      archiveReader,
				Prelude: &archive.Prelude{},
			}
		}
		err = restore.archive.Prelude.Read(restore.archive.In)
		if err != nil {
			return err
		}
		log.Logvf(log.DebugLow, `archive format version "%v"`, restore.archive.Prelude.Header.FormatVersion)
		log.Logvf(log.DebugLow, `archive server version "%v"`, restore.archive.Prelude.Header.ServerVersion)
		log.Logvf(log.DebugLow, `archive tool version "%v"`, restore.archive.Prelude.Header.ToolVersion)
		target, err = restore.archive.Prelude.NewPreludeExplorer()
		if err != nil {
			return err
		}
	} else if restore.TargetDirectory != "-" {
		var usedDefaultTarget bool
		if restore.TargetDirectory == "" {
			restore.TargetDirectory = "dump"
			log.Logv(log.Always, "using default 'dump' directory")
			usedDefaultTarget = true
		}
		target, err = newActualPath(restore.TargetDirectory)
		if err != nil {
			if usedDefaultTarget {
				log.Logv(log.Always, "see mongorestore --help for usage information")
			}
			return fmt.Errorf("mongorestore target '%v' invalid: %v", restore.TargetDirectory, err)
		}
		// handle cases where the user passes in a file instead of a directory
		if !target.IsDir() {
			log.Logv(log.DebugLow, "mongorestore target is a file, not a directory")
			err = restore.handleBSONInsteadOfDirectory(restore.TargetDirectory)
			if err != nil {
				return err
			}
		} else {
			log.Logv(log.DebugLow, "mongorestore target is a directory, not a file")
		}
	}
	if restore.NSOptions.Collection != "" &&
		restore.OutputOptions.NumParallelCollections > 1 &&
		restore.OutputOptions.NumInsertionWorkers == 1 {
		// handle special parallelization case when we are only restoring one collection
		// by mapping -j to insertion workers rather than parallel collections
		log.Logvf(log.DebugHigh,
			"setting number of insertions workers to number of parallel collections (%v)",
			restore.OutputOptions.NumParallelCollections)
		restore.OutputOptions.NumInsertionWorkers = restore.OutputOptions.NumParallelCollections
	}
	if restore.InputOptions.Archive != "" {
		if int(restore.archive.Prelude.Header.ConcurrentCollections) > restore.OutputOptions.NumParallelCollections {
			restore.OutputOptions.NumParallelCollections = int(restore.archive.Prelude.Header.ConcurrentCollections)
			restore.OutputOptions.NumInsertionWorkers = int(restore.archive.Prelude.Header.ConcurrentCollections)
			log.Logvf(log.Always,
				"setting number of parallel collections to number of parallel collections in archive (%v)",
				restore.archive.Prelude.Header.ConcurrentCollections,
			)
		}
	}

	// Create the demux before intent creation, because muted archive intents need
	// to register themselves with the demux directly
	if restore.InputOptions.Archive != "" {
		restore.archive.Demux = archive.CreateDemux(restore.archive.Prelude.NamespaceMetadatas, restore.archive.In)
	}

	switch {
	case restore.InputOptions.Archive != "":
		log.Logvf(log.Always, "preparing collections to restore from")
		err = restore.CreateAllIntents(target)
	case restore.NSOptions.DB != "" && restore.NSOptions.Collection == "":
		log.Logvf(log.Always,
			"building a list of collections to restore from %v dir",
			target.Path())
		err = restore.CreateIntentsForDB(
			restore.NSOptions.DB,
			target,
		)
	case restore.NSOptions.DB != "" && restore.NSOptions.Collection != "" && restore.TargetDirectory == "-":
		log.Logvf(log.Always, "setting up a collection to be read from standard input")
		err = restore.CreateStdinIntentForCollection(
			restore.NSOptions.DB,
			restore.NSOptions.Collection,
		)
	case restore.NSOptions.DB != "" && restore.NSOptions.Collection != "":
		log.Logvf(log.Always, "checking for collection data in %v", target.Path())
		err = restore.CreateIntentForCollection(
			restore.NSOptions.DB,
			restore.NSOptions.Collection,
			target,
		)
	default:
		log.Logvf(log.Always, "preparing collections to restore from")
		err = restore.CreateAllIntents(target)
	}
	if err != nil {
		return fmt.Errorf("error scanning filesystem: %v", err)
	}

	if restore.isMongos && restore.manager.HasConfigDBIntent() && restore.NSOptions.DB == "" {
		return fmt.Errorf("cannot do a full restore on a sharded system - " +
			"remove the 'config' directory from the dump directory first")
	}

	if restore.InputOptions.OplogFile != "" {
		err = restore.CreateIntentForOplog()
		if err != nil {
			return fmt.Errorf("error reading oplog file: %v", err)
		}
	}
	if restore.InputOptions.OplogReplay && restore.manager.Oplog() == nil {
		return fmt.Errorf("no oplog file to replay; make sure you run mongodump with --oplog")
	}
	if restore.manager.GetOplogConflict() {
		return fmt.Errorf("cannot provide both an oplog.bson file and an oplog file with --oplogFile, " +
			"nor can you provide both a local/oplog.rs.bson and a local/oplog.$main.bson file.")
	}

	conflicts := restore.manager.GetDestinationConflicts()
	if len(conflicts) > 0 {
		for _, conflict := range conflicts {
			log.Logvf(log.Always, "%s", conflict.Error())
		}
		return fmt.Errorf("cannot restore with conflicting namespace destinations")
	}

	if restore.OutputOptions.DryRun {
		log.Logvf(log.Always, "dry run completed")
		return nil
	}

	demuxFinished := make(chan interface{})
	var demuxErr error
	if restore.InputOptions.Archive != "" {
		namespaceChan := make(chan string, 1)
		namespaceErrorChan := make(chan error)
		restore.archive.Demux.NamespaceChan = namespaceChan
		restore.archive.Demux.NamespaceErrorChan = namespaceErrorChan

		go func() {
			demuxErr = restore.archive.Demux.Run()
			close(demuxFinished)
		}()
		// consume the new namespace announcement from the demux for all of the special collections
		// that get cached when being read out of the archive.
		// The first regular collection found gets pushed back on to the namespaceChan
		// consume the new namespace announcement from the demux for all of the collections that get cached
		for {
			ns, ok := <-namespaceChan
			// the archive can have only special collections. In that case we keep reading until
			// the namespaces are exhausted, indicated by the namespaceChan being closed.
			if !ok {
				break
			}
			intent := restore.manager.IntentForNamespace(ns)
			if intent == nil {
				return fmt.Errorf("no intent for collection in archive: %v", ns)
			}
			if intent.IsSystemIndexes() ||
				intent.IsUsers() ||
				intent.IsRoles() ||
				intent.IsAuthVersion() {
				log.Logvf(log.DebugLow, "special collection %v found", ns)
				namespaceErrorChan <- nil
			} else {
				// Put the ns back on the announcement chan so that the
				// demultiplexer can start correctly
				log.Logvf(log.DebugLow, "first non special collection %v found."+
					" The demultiplexer will handle it and the remainder", ns)
				namespaceChan <- ns
				break
			}
		}
	}

	// If restoring users and roles, make sure we validate auth versions
	if restore.ShouldRestoreUsersAndRoles() {
		log.Logv(log.Info, "comparing auth version of the dump directory and target server")
		restore.authVersions.Dump, err = restore.GetDumpAuthVersion()
		if err != nil {
			return fmt.Errorf("error getting auth version from dump: %v", err)
		}
		restore.authVersions.Server, err = auth.GetAuthVersion(restore.SessionProvider)
		if err != nil {
			return fmt.Errorf("error getting auth version of server: %v", err)
		}
		err = restore.ValidateAuthVersions()
		if err != nil {
			return fmt.Errorf(
				"the users and roles collections in the dump have an incompatible auth version with target server: %v",
				err)
		}
	}

	err = restore.LoadIndexesFromBSON()
	if err != nil {
		return fmt.Errorf("restore error: %v", err)
	}

	// Restore the regular collections
	if restore.InputOptions.Archive != "" {
		restore.manager.UsePrioritizer(restore.archive.Demux.NewPrioritizer(restore.manager))
	} else if restore.OutputOptions.NumParallelCollections > 1 {
		restore.manager.Finalize(intents.MultiDatabaseLTF)
	} else {
		// use legacy restoration order if we are single-threaded
		restore.manager.Finalize(intents.Legacy)
	}

	restore.termChan = make(chan struct{})

	if err := restore.RestoreIntents(); err != nil {
		return err
	}

	// Restore users/roles
	if restore.ShouldRestoreUsersAndRoles() {
		err = restore.RestoreUsersOrRoles(restore.manager.Users(), restore.manager.Roles())
		if err != nil {
			return fmt.Errorf("restore error: %v", err)
		}
	}

	// Restore oplog
	if restore.InputOptions.OplogReplay {
		err = restore.RestoreOplog()
		if err != nil {
			return fmt.Errorf("restore error: %v", err)
		}
	}

	defer log.Logv(log.Always, "done")

	if restore.InputOptions.Archive != "" {
		<-demuxFinished
		return demuxErr
	}

	return nil
}

func (restore *MongoRestore) getArchiveReader() (rc io.ReadCloser, err error) {
	if restore.InputOptions.Archive == "-" {
		rc = ioutil.NopCloser(restore.InputReader)
	} else {
		targetStat, err := os.Stat(restore.InputOptions.Archive)
		if err != nil {
			return nil, err
		}
		if targetStat.IsDir() {
			defaultArchiveFilePath := filepath.Join(restore.InputOptions.Archive, "archive")
			if restore.InputOptions.Gzip {
				defaultArchiveFilePath = defaultArchiveFilePath + ".gz"
			}
			rc, err = os.Open(defaultArchiveFilePath)
			if err != nil {
				return nil, err
			}
		} else {
			rc, err = os.Open(restore.InputOptions.Archive)
			if err != nil {
				return nil, err
			}
		}
	}
	if restore.InputOptions.Gzip {
		gzrc, err := gzip.NewReader(rc)
		if err != nil {
			return nil, err
		}
		return &util.WrappedReadCloser{gzrc, rc}, nil
	}
	return rc, nil
}

func (restore *MongoRestore) HandleInterrupt() {
	if restore.termChan != nil {
		close(restore.termChan)
	}
}
